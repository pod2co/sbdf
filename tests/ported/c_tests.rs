use sbdf::{BoolArray, ColumnProperties, ColumnSlice, EncodedBitArray, EncodedValue, Object, Sbdf};
use std::{fs::File, io::Read};

fn read_file(file_name: &str) -> Sbdf {
    let mut file = File::open(file_name).unwrap();
    let mut bytes = Vec::new();
    file.read_to_end(&mut bytes).unwrap();
    Sbdf::from_bytes(&bytes).unwrap()
}

macro_rules! test_snapshot {
    ($name:ident, $file:expr) => {
        #[test]
        fn $name() {
            let sbdf = read_file($file);
            insta::assert_ron_snapshot!(sbdf);
        }
    };
}

test_snapshot!(
    sample_0_binary,
    "tests/ported/spotfire-sbdf-c/tests/samples/0_Binary.sbdf"
);
test_snapshot!(
    sample_0_boolean,
    "tests/ported/spotfire-sbdf-c/tests/samples/0_Boolean.sbdf"
);
test_snapshot!(
    sample_0_date,
    "tests/ported/spotfire-sbdf-c/tests/samples/0_Date.sbdf"
);
test_snapshot!(
    sample_0_datetime,
    "tests/ported/spotfire-sbdf-c/tests/samples/0_DateTime.sbdf"
);
test_snapshot!(
    sample_0_decimal,
    "tests/ported/spotfire-sbdf-c/tests/samples/0_Decimal.sbdf"
);
test_snapshot!(
    sample_0_double,
    "tests/ported/spotfire-sbdf-c/tests/samples/0_Double.sbdf"
);
test_snapshot!(
    sample_0_float,
    "tests/ported/spotfire-sbdf-c/tests/samples/0_Float.sbdf"
);
test_snapshot!(
    sample_0_integer,
    "tests/ported/spotfire-sbdf-c/tests/samples/0_Integer.sbdf"
);
test_snapshot!(
    sample_0_long,
    "tests/ported/spotfire-sbdf-c/tests/samples/0_Long.sbdf"
);
test_snapshot!(
    sample_0_string,
    "tests/ported/spotfire-sbdf-c/tests/samples/0_String.sbdf"
);
test_snapshot!(
    sample_0_time,
    "tests/ported/spotfire-sbdf-c/tests/samples/0_Time.sbdf"
);
test_snapshot!(
    sample_0_timespan,
    "tests/ported/spotfire-sbdf-c/tests/samples/0_TimeSpan.sbdf"
);
test_snapshot!(
    sample_0,
    "tests/ported/spotfire-sbdf-c/tests/samples/0.sbdf"
);
test_snapshot!(
    sample_1_binary,
    "tests/ported/spotfire-sbdf-c/tests/samples/1_Binary.sbdf"
);
test_snapshot!(
    sample_1_boolean,
    "tests/ported/spotfire-sbdf-c/tests/samples/1_Boolean.sbdf"
);
test_snapshot!(
    sample_1_date,
    "tests/ported/spotfire-sbdf-c/tests/samples/1_Date.sbdf"
);
test_snapshot!(
    sample_1_datetime,
    "tests/ported/spotfire-sbdf-c/tests/samples/1_DateTime.sbdf"
);
test_snapshot!(
    sample_1_decimal,
    "tests/ported/spotfire-sbdf-c/tests/samples/1_Decimal.sbdf"
);
test_snapshot!(
    sample_1_double,
    "tests/ported/spotfire-sbdf-c/tests/samples/1_Double.sbdf"
);
test_snapshot!(
    sample_1_float,
    "tests/ported/spotfire-sbdf-c/tests/samples/1_Float.sbdf"
);
test_snapshot!(
    sample_1_integer,
    "tests/ported/spotfire-sbdf-c/tests/samples/1_Integer.sbdf"
);
test_snapshot!(
    sample_1_long,
    "tests/ported/spotfire-sbdf-c/tests/samples/1_Long.sbdf"
);
test_snapshot!(
    sample_1_string,
    "tests/ported/spotfire-sbdf-c/tests/samples/1_String.sbdf"
);
test_snapshot!(
    sample_1_time,
    "tests/ported/spotfire-sbdf-c/tests/samples/1_Time.sbdf"
);
test_snapshot!(
    sample_1_timespan,
    "tests/ported/spotfire-sbdf-c/tests/samples/1_TimeSpan.sbdf"
);
test_snapshot!(
    sample_1,
    "tests/ported/spotfire-sbdf-c/tests/samples/1.sbdf"
);
test_snapshot!(
    sample_100_binary,
    "tests/ported/spotfire-sbdf-c/tests/samples/100_Binary.sbdf"
);
test_snapshot!(
    sample_100_boolean,
    "tests/ported/spotfire-sbdf-c/tests/samples/100_Boolean.sbdf"
);
test_snapshot!(
    sample_100_date,
    "tests/ported/spotfire-sbdf-c/tests/samples/100_Date.sbdf"
);
test_snapshot!(
    sample_100_datetime,
    "tests/ported/spotfire-sbdf-c/tests/samples/100_DateTime.sbdf"
);
test_snapshot!(
    sample_100_decimal,
    "tests/ported/spotfire-sbdf-c/tests/samples/100_Decimal.sbdf"
);
test_snapshot!(
    sample_100_double,
    "tests/ported/spotfire-sbdf-c/tests/samples/100_Double.sbdf"
);
test_snapshot!(
    sample_100_float,
    "tests/ported/spotfire-sbdf-c/tests/samples/100_Float.sbdf"
);
test_snapshot!(
    sample_100_integer,
    "tests/ported/spotfire-sbdf-c/tests/samples/100_Integer.sbdf"
);
test_snapshot!(
    sample_100_long,
    "tests/ported/spotfire-sbdf-c/tests/samples/100_Long.sbdf"
);
test_snapshot!(
    sample_100_string,
    "tests/ported/spotfire-sbdf-c/tests/samples/100_String.sbdf"
);
test_snapshot!(
    sample_100_time,
    "tests/ported/spotfire-sbdf-c/tests/samples/100_Time.sbdf"
);
test_snapshot!(
    sample_100_timespan,
    "tests/ported/spotfire-sbdf-c/tests/samples/100_TimeSpan.sbdf"
);
test_snapshot!(
    sample_100,
    "tests/ported/spotfire-sbdf-c/tests/samples/100.sbdf"
);
test_snapshot!(
    sample_no_columns_binary,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns_Binary.sbdf"
);
test_snapshot!(
    sample_no_columns_boolean,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns_Boolean.sbdf"
);
test_snapshot!(
    sample_no_columns_date,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns_Date.sbdf"
);
test_snapshot!(
    sample_no_columns_datetime,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns_DateTime.sbdf"
);
test_snapshot!(
    sample_no_columns_decimal,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns_Decimal.sbdf"
);
test_snapshot!(
    sample_no_columns_double,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns_Double.sbdf"
);
test_snapshot!(
    sample_no_columns_float,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns_Float.sbdf"
);
test_snapshot!(
    sample_no_columns_integer,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns_Integer.sbdf"
);
test_snapshot!(
    sample_no_columns_long,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns_Long.sbdf"
);
test_snapshot!(
    sample_no_columns_string,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns_String.sbdf"
);
test_snapshot!(
    sample_no_columns_time,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns_Time.sbdf"
);
test_snapshot!(
    sample_no_columns_timespan,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns_TimeSpan.sbdf"
);
test_snapshot!(
    sample_no_columns,
    "tests/ported/spotfire-sbdf-c/tests/samples/no_columns.sbdf"
);
test_snapshot!(
    root_2x4_cs,
    "tests/ported/spotfire-sbdf-c/tests/2x4_cs.sbdf"
);
test_snapshot!(
    root_3x3_cs,
    "tests/ported/spotfire-sbdf-c/tests/2x4_cs.sbdf"
);

/// We don't want to snapshot certain samples because they are too large.
///
/// Instead we'll just make sure we can read the file without an error.
macro_rules! test_no_snapshot {
    ($name:ident, $file:expr) => {
        #[test]
        fn $name() {
            let _sbdf = read_file($file);
        }
    };
}
test_no_snapshot!(
    sample_10001_binary,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001_Binary.sbdf"
);
test_no_snapshot!(
    sample_10001_boolean,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001_Boolean.sbdf"
);
test_no_snapshot!(
    sample_10001_date,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001_Date.sbdf"
);
test_no_snapshot!(
    sample_10001_datetime,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001_DateTime.sbdf"
);
test_no_snapshot!(
    sample_10001_decimal,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001_Decimal.sbdf"
);
test_no_snapshot!(
    sample_10001_double,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001_Double.sbdf"
);
test_no_snapshot!(
    sample_10001_float,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001_Float.sbdf"
);
test_no_snapshot!(
    sample_10001_integer,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001_Integer.sbdf"
);
test_no_snapshot!(
    sample_10001_long,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001_Long.sbdf"
);
test_no_snapshot!(
    sample_10001_string,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001_String.sbdf"
);
test_no_snapshot!(
    sample_10001_time,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001_Time.sbdf"
);
test_no_snapshot!(
    sample_10001_timespan,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001_TimeSpan.sbdf"
);
test_no_snapshot!(
    sample_10001,
    "tests/ported/spotfire-sbdf-c/tests/samples/10001.sbdf"
);

#[test]
fn loads_values() {
    let sbdf = read_file("tests/ported/spotfire-sbdf-c/tests/3x3_cs.sbdf");
    let all_values = sbdf
        .table_slices
        .iter()
        .map(|table_slice| {
            table_slice
                .column_slices
                .iter()
                .map(|column| column.load_values().unwrap())
                .collect::<Vec<_>>()
        })
        .collect::<Vec<_>>();
    insta::assert_ron_snapshot!((&sbdf, all_values));
}

#[test]
fn decodes_bit_arrays() {
    let test1 = (
        EncodedValue::BitArray(EncodedBitArray {
            bit_count: 8,
            bytes: vec![38].into_boxed_slice(),
        }),
        vec![false, false, true, false, false, true, true, false].into_boxed_slice(),
    );
    let test2 = (
        EncodedValue::BitArray(EncodedBitArray {
            bit_count: 10,
            bytes: vec![147, 64].into_boxed_slice(),
        }),
        vec![
            true, false, false, true, false, false, true, true, false, true,
        ]
        .into_boxed_slice(),
    );
    let cases = vec![test1, test2];

    for (values, expected) in cases {
        let slice = ColumnSlice {
            values,
            properties: ColumnProperties::default(),
        };

        let result = slice.load_values();

        match result.unwrap().as_ref() {
            Object::BoolArray(BoolArray(ba)) => {
                assert_eq!(ba, &expected);
            }
            _ => panic!("expected bool array"),
        }
    }
}
