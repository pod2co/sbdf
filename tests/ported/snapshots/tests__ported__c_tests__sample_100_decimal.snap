---
source: tests/ported/c_tests.rs
expression: sbdf
---
Sbdf(
  file_header: FileHeader(
    major_version: 1,
    minor_version: 0,
  ),
  table_metadata: TableMetadata(
    metadata: [
      Metadata(
        name: "SbdfTest.TableMetaDecimal",
        value: Decimal((119, 26, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48)),
      ),
    ],
    columns: [
      ColumnMetadata(
        name: "Decimal",
        ty: Decimal,
        other: [
          Metadata(
            name: "SbdfTest.ColumnMetaDecimal0",
            value: Decimal((119, 26, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48)),
          ),
        ],
      ),
    ],
  ),
  table_slices: [
    TableSlice(
      column_slices: [
        ColumnSlice(
          values: Plain(
            value: DecimalArray([
              (87, 105, 168, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (251, 89, 17, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (70, 95, 18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (51, 211, 188, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (33, 8, 199, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (150, 10, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 60, 48),
              (15, 61, 209, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (39, 111, 21, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (253, 113, 219, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (235, 166, 229, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (189, 121, 23, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (217, 219, 239, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (8, 127, 24, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (199, 16, 250, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (83, 132, 25, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (158, 137, 26, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (163, 122, 14, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (233, 142, 27, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (145, 175, 24, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (52, 148, 28, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (127, 153, 29, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (109, 25, 45, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (91, 78, 55, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (21, 164, 31, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (73, 131, 65, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (96, 169, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (55, 184, 75, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (171, 174, 33, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (37, 237, 85, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 0, 0, 64, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (65, 185, 35, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (1, 87, 106, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (140, 190, 36, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (239, 139, 116, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (215, 195, 37, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (221, 192, 126, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (203, 245, 136, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (109, 206, 39, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (185, 42, 147, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (44, 21, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 60, 48),
              (167, 95, 157, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (3, 217, 41, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (149, 148, 167, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (131, 201, 177, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (153, 227, 43, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (113, 254, 187, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (228, 232, 44, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (95, 51, 198, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (47, 238, 45, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (77, 104, 208, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (59, 157, 218, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (197, 248, 47, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (16, 254, 48, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (23, 7, 239, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (91, 3, 50, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (5, 60, 249, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (119, 26, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 60, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (241, 13, 52, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (225, 165, 13, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (60, 19, 53, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (207, 218, 23, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (135, 24, 54, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 128, 0, 0, 64, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (171, 68, 44, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (29, 35, 56, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (153, 121, 54, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (104, 40, 57, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (135, 174, 64, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (117, 227, 74, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (254, 50, 59, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (73, 56, 60, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (81, 77, 95, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (194, 31, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 60, 48),
              (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 64, 48),
              (63, 130, 105, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
              (223, 66, 62, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 58, 48),
              (45, 183, 115, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 56, 48),
            ]),
          ),
          properties: [
            Property(
              name: "IsInvalid",
              values: BitArray(
                bit_count: 100,
                bytes: [
                  4,
                  8,
                  16,
                  32,
                  64,
                  128,
                  2,
                  4,
                  8,
                  16,
                  32,
                  64,
                  128,
                ],
              ),
            ),
            Property(
              name: "ErrorCode",
              values: Plain(
                value: StringArray([
                  "",
                  "",
                  "Error 35",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "Error 44",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "Error 53",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "Error 62",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "Error 71",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "Error 80",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "Error 89",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "Error 98",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "Error 107",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "Error 116",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "Error 125",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                  "",
                ]),
              ),
            ),
            Property(
              name: "HasReplacedValue",
              values: BitArray(
                bit_count: 100,
                bytes: [
                  1,
                  0,
                  0,
                  1,
                  0,
                  16,
                  1,
                  0,
                  16,
                  1,
                  0,
                  16,
                  0,
                ],
              ),
            ),
          ],
        ),
      ],
    ),
  ],
)
